tests/php/Storage/AssetStoreTest.php

```diff
@@ -520,4 +621,68 @@ public function testProtect()
             $backend->getVisibility($fishTuple['Filename'], $fishTuple['Hash'])
         );
     }
+
+    public function testRename()
+    {
+        $backend = $this->getBackend();
+        $fish1 = realpath(__DIR__ . '/../ImageTest/test-image-high-quality.jpg');
+
+        // Create file with various variants
+        $fish1Tuple = $backend->setFromLocalFile($fish1, 'directory/lovely-fish.jpg');
+        $backend->setFromLocalFile($fish1, $fish1Tuple['Filename'], $fish1Tuple['Hash'], 'somevariant');
+        $backend->setFromLocalFile($fish1, $fish1Tuple['Filename'], $fish1Tuple['Hash'], 'anothervariant');
+
+        // Move to new filename
+        $newFilename = 'another-file.jpg';
+        $confirmedFilename = $backend->rename($fish1Tuple['Filename'], $fish1Tuple['Hash'], $newFilename);
+
+        // Check result
+        $this->assertEquals($newFilename, $confirmedFilename);
+        $this->assertNotEquals($fish1Tuple['Filename'], $confirmedFilename);
+
+        // Check old files no longer exist
+        $this->assertFalse($backend->exists($fish1Tuple['Filename'], $fish1Tuple['Hash']));
+        $this->assertFalse($backend->exists($fish1Tuple['Filename'], $fish1Tuple['Hash'], 'somevariant'));
+        $this->assertFalse($backend->exists($fish1Tuple['Filename'], $fish1Tuple['Hash'], 'anothervariant'));
+
+        // New files exist
+        $this->assertTrue($backend->exists($newFilename, $fish1Tuple['Hash']));
+        $this->assertTrue($backend->exists($newFilename, $fish1Tuple['Hash'], 'somevariant'));
+        $this->assertTrue($backend->exists($newFilename, $fish1Tuple['Hash'], 'anothervariant'));
+
+        // Ensure we aren't getting false positives for exists()
+        $this->assertFalse($backend->exists($fish1Tuple['Filename'], $fish1Tuple['Hash'], 'nonvariant'));
+        $this->assertFalse($backend->exists($fish1Tuple['Filename'], sha1('nothash')));
+        $this->assertFalse($backend->exists($newFilename, $fish1Tuple['Hash'], 'nonvariant'));
+        $this->assertFalse($backend->exists('notfilename.jpg', $fish1Tuple['Hash']));
```

Comment:
The point is to test that the moved file doesn't cause the false positives. It's meant to be a part of this test.